### DISCUSSION TITLE
Assigning null to T? if typeparameter T is unconstrained

### DISCUSSION THREAD

**Discussion by PaulRitter** [dotnet/csharplang#5911] [Q&A] [answered]
Excuse me if this has already been asked before, i searched through the discussions & issues and didn't find this question.

I've been working with generics paired with nullability and one pitfall i've ran into again and again is the following:
These methods are fine:
```csharp
T? ReturnVal<T>() where T : class => null;
```
```csharp
T? ReturnVal<T>() where T : struct => null;
```

While this one produces an error, claiming "Cannot convert null to type parameter 'T' because it could be a non-nullable value type"
```csharp
T? ReturnVal<T>() => null;
```
Now we are obviously trying to return `T?` here, not `T`. However, for this, we have an explanation:

Reading through the spec, it seems that `T?` is not allowed for unconstrained type parameters.
https://github.com/dotnet/csharplang/blob/65580d88f2074df48be58d2069aa24f982384a83/proposals/csharp-8.0/nullable-reference-types-specification.md?plain=1#L163
As far as i can tell, this causes `T?` in our case to be implicitly converted to `T`. Additionally, this causes substituting `null` with `default(T?)` to return `default(T)` instead of `null`, meaning the method will not return null values when called with a struct as type parameter.

Now my question, at the risk of sounding a bit stupid, is: Why? As far as i can tell, there isn't really any type i could pass as a type parameter that could not be made nullable, so why impose this restriction? Is this due to internal limits since struct-null-values are actually the default of Nullable<T>, not genuine null-pointers?

Genuinely curious what could be the cause of this, thanks in advance :)
\- Paul

**Comment by bernd5** [comment, marked as answer]
@PaulRitter the problem is that not the JITTer implements nullability - it is done by the IL Compiler.

For example: `T? ReturnVal<T>() where T : struct => null;`
is compiled to:

```csharp
static Nullable<T> ReturnVal<T>() where T : struct
{
        return null;
}
```

while `T? ReturnVal<T>() where T : class => null;`
is compiled to:
```csharp
static T ReturnVal<T>() where T : class
{
        return null;
}
````

**Reply by PaulRitter** [reply]
Makes sense, i figured as much. Thanks for the quick response!
Although this begs the question if there may/should be some way to handle this issue, since on the developer-facing side, if you ignore the constraints imposed by the nullability implementation, it *should* be possible.

For now tho, i'll just continue to specialcase value- & referencetypes :/

**Reply by Joe4evr** [reply]
This has been asked about many times before and the simple answer is that it's not possible to combine the two scenarios, as there is no way to emit IL that is both `T` for reference types and `Nullable<T>` for value types for the same symbol.

**Comment by TahirAhmadov** [comment]
Please look at this: https://github.com/dotnet/csharplang/discussions/4080
There is a way to make it work, but unfortunately it seems the language went in the direction of sweeping this under the rug.

**Reply by CyrusNajmabadi** [reply]
It was not swept under the rug.  As Fred mentioned designs very similar to yours were considered. They were rejected because their negatives were felt to far outweigh the positives. 

Sometimes (actually most of the time) there are zero paths that are completely positive. Instead your have many paths with disparate tradeoffs.  Assessing that and making the best educated decisions on what the right set of tradeoffs are is the exact opposite of sweeping under the rug.

**Reply by TahirAhmadov** [reply]
I know what you are saying is true in most cases, but I think my proposed solution for it had no real downsides - except for amount of effort. And I totally understand that concern, too. What peeves me is the "nail in the coffin" of using (wasting, IMO) the `T?` syntax on the "default" value approach (zero for values, null for classes). Fine, it's not a priority to do the expensive solution which solves everything, but surely there was no need to make it darn near impossible to ever do it in the future.

**Reply by CyrusNajmabadi** [reply]
> I know what you are saying is true in most cases, but I think my proposed solution for it had no real downsides 

You're basically saying: if we were able to make everything work such that classes and structs could be unified at a core representation level, then we'd have a language and runtime where classes and structs were unified at a core representation level.   This is circular.  Yes, if we can make everything work, then we could make everything work.  But the point is that we could see no actual path to make everything work that was in any way feasible at both an engineering level or business level in any sort of acceptable user timeframe. 

The point made was effectively there was no realistic path to that, and it was akin to having a whole new runtime built on an entirely different system.

Having everything be rebuilt is not realistic, and framing people assessing that and making the decision to keep compat and keep all these investments of decades is not "sweeping under the rug", it's people saying: it's better to have a solution now that does well in 90% of cases rather than burning to the ground and starting over to get 100% a decade from now with massive cost to the entire ecosystem.

> but surely there was no need to make it darn near impossible to ever do it in the future.

I don't see why it would be impossible in the future, just enormously unlikely because of the cost.  At the point that we would be making such a deep rebuilding of the runtime, we'd just change the language as well.    Basically, if things are going to change and break that much, then just assume that everything is on the table in terms of what we do going forward as being beholden to the past is clearly not part of the equation anymore.

> And I totally understand that concern, too.

Then why describe it as "sweeping under the rug"?  People doing a ton of investigation and discovering that the costs are just astronomical is not sweeping under the rug.  It's a solid engineering effort that comes up with options which are then assessed to see how acceptable that is is the exact opposite.  No one was trying to hide anything.

**Reply by TahirAhmadov** [reply]
OK I think you grossly misunderstood my proposal at #4080. It would not entail any rebuilding of anything; just copying and pasting some of what is already done for `Nullable<T>` to a new type, `UniversalNullable<T>`, and compiling code which uses unconstrained `T?` to use that new type. Yes, that's more expensive that the "default" approach, but it's nowhere near what you described. I'm not crazy, after all :)

**Reply by CyrusNajmabadi** [reply]
> What peeves me is the "nail in the coffin" of using (wasting, IMO) the T? syntax on the "default" value approach

We strongly considered and debated this exact point at length.  Our choice was either:

1. to reserve this syntax toward the speculative future that seems enormously unlikely.  indeed, such a future was felt to be so unlikely that if it did come, it would certainly come with a new lang, so reservation of this syntax was not necessary.  Come up with some other syntax that would likely be bizarre and hard to explain in its own right.
2. use this syntax.   have a reasonable syntactic solution for people that would serve the needs of this space immediately, and likely for the next 10+ years.

You're asking us to not be pragmatic here.  We don't make these decisions unaware of these topics.  We're just not willing to handcuff ourselves to a highly improbable far-off future.  Not when it harms the very real present and future that we know we can get to.

**Reply by CyrusNajmabadi** [reply]
> OK I think you grossly misunderstood my proposal at https://github.com/dotnet/csharplang/discussions/4080. 

I'm not misunderstanding.  This was considered and found to be extremely expensive.  

> It would not entail any rebuilding of anything; just copying and pasting some of what is already done for `Nullable<T>` to a new type

That's not what it would entail.  To *begin* with, that belies a deep misunderstanding of this problem space.  `Nullable<T>`, in particular, can work at the runtime level *precisely* because it knows it knows that it only wraps a struct, and that a `Nullable<T>` is not actually a `T`.  So if you copy/paste that and attempt to make that work with reference types you run into a heap (har har) of problems.  Interfaces don't actually work here.  Codegen explodes for reference types.  etc. etc.

Again, you're handwaving and saying: if you just solve this, then you've solved it.

The point was made that this is absolutely not at all easy or cheap to do.  And they also tend to still have holes in them as well. This same topic has come up with the runtime as we discuss roles/shapes, and even in that targeted *new* area, it's still fraught with challenges, where even as we see waht the runtime does/doesn't find feasible, this still can potentially have type-holes that are still problematic and just as observable as the `T?` issue between value types and reference types.

**Reply by TahirAhmadov** [reply]
> That's not what it would entail. To _begin_ with, that belies a deep misunderstanding of this problem space. `Nullable<T>`, in particular, can work at the runtime level _precisely_ because it knows it knows that it only wraps a struct, and that a `Nullable<T>` is not actually a `T`. So if you copy/paste that and attempt to make that work with reference types you run into a heap (har har) of problems. Interfaces don't actually work here. Codegen explodes for reference types. etc. etc.

Regarding interfaces, when `Nullable<T>` is boxed, just `T` is boxed, not `Nullable<T>`. Same idea with wrapping a class; if you try boxing a `UniversalNullable<T>`, and `T` is a struct, then `T` is boxed; if `T` is a class, then that reference is just returned - essentially a noop. I'm not saying it's an exact copy from the `Nullable<T>` underlying code, but it's similar - with some changes. In both cases, interfaces continue working just fine.
I recall Fred said it was deemed to difficult to expand `Nullable<T>` to support this. And I agree! A new type would have been easier to accomplish.
I'm not sure what you mean by "codegen explodes".

**Reply by TahirAhmadov** [reply]
To underscore exactly how feasible (not saying easy, but feasible) it is, I actually implemented it. If you go to #4080, and scroll to the bottom, I actually implemented it myself. Unfortunately, I had to do a whole bunch of things manually due to lack of language support, but none of that code is crazy complicated.

**Reply by CyrusNajmabadi** [reply]
> Unfortunately, I had to do a whole bunch of things manually due to lack of language support, but none of that code is crazy complicated.

...

This is exactly handwaving.  Yes, if you implement it *but have gaps in the impl* then it's not a problem. But if you want the system to be understood end to end, then it's not trivial.

> I'm not sure what you mean by "codegen explodes".
> I actually implemented it myself. 

With your impl, if you actually use it, and you end up instantiating your Null type with things like string/object/etc. you will get codegen explosion.  For example, you will share no jit code between `List<Null<object>>` and `List<Null<string>>` .  

This is again saying: if you take everything that doesn't work and make it work, then it woudl work.

Furthermore, if you had this, you would immediately run into type system problems.  For example, say you had this:

```c#
class C<T> where T : class
{
    public static List<T?> List = new();
}

class D<T>
{
    public static List<T?> List = new();
}

List<string?> list1 = C<string>.List;
List<string?> list2 = D<string>.List;
```

How do you get this to work in your world?  These need to be the same types as per how the language and runtime work *today*.  However, in your world, `D<string>.List` is actually a `List<Null<string>>` which is an entirely different type than `List<string>`.

**Reply by CyrusNajmabadi** [reply]
The same holds true for `where T : struct`.  In one case you'll have `Nullable<T>` and the other `Null<T>`.  Your system won't type unify and that will itself expose in varying ways.

**Reply by TahirAhmadov** [reply]
> This is exactly handwaving. Yes, if you implement it _but have gaps in the impl_ then it's not a problem. But if you want the system to be understood end to end, then it's not trivial.

The gaps were caused by the fact that it's a type I created in user code; these gaps would not have existed if it was in the framework/language.

> With your impl, if you actually use it, and you end up instantiating your Null type with things like string/object/etc. you will get codegen explosion. For example, you will share no jit code between `List<Null<object>>` and `List<Null<string>>` .

I'm not sure I understand. Aren't all references the same size? How is this different than any generic type out there?

> Furthermore, if you had this, you would immediately run into type system problems. For example, say you had this:
> 
> ```cs
> class C<T> where T : class
> {
>     public static List<string> List = new();
> }
> 
> class D<T>
> {
>     public static List<T?> List = new();
> }
> 
> List<string> list1 = C<string>.List;
> List<string> list2 = D<string>.List;
> ```
> 
> How do you get this to work in your world? These need to be the same types as per how the language and runtime work _today_. However, in your world, `D<string>.List` is actually a `List<Null<string>>` which is an entirely different type than `List<string>`.

We're comparing to how everything worked before `T?` was made to mean "default", not "after". "Before", that `D<T>` class would not compile, because `T?` was not allowed, right? So, if `D<string>.List` means `List<Null<string>>`, it doesn't break anything - compared to that previous version. 
Now, on the language side, say you have:
```cs
var list2 = D<string>.List;
list2.Add("abc"); // implicit wrapping into Null<string>
string k = list2[0]!; // implicit un-wrapping into string, and ! null-suppresses the warning
```

**Reply by CyrusNajmabadi** [reply]
> The gaps were caused by the fact that it's a type I created in user code; these gaps would not have existed if it was in the framework/language.

...

Yes, this is exactly: if you manage to solve the entire problem here, then the problem is solved.  THe point is that we have no known way to go about solving this effectively.  

> I'm not sure I understand. Aren't all references the same size? How is this different than any generic type out there?

The runtime doesn't base it on size.  That's not how the jit works here.

> "Before", that D<T> class would not compile, because T? was not allowed, right? So, if D<string>.List means List<Null<string>>, it doesn't break anything -

I mean, it doesn't break anything in the sense that it would have new semantics as you define them.  But it would break htings in that `List<string>` and `List<string?>` would not be the same types, which would be highly undesirable (and basically go against entirely the direction we went with NRT).  

If you're suggesting that we take a generics path with null such that `string` and `string?` are not binary compatible, then that woudl be a complete non-starter on its own.

**Reply by CyrusNajmabadi** [reply]
@TahirAhmadov I'm getting your position coming down to effectively that you want a strong typesystem distinction between `T?` and `T`.  If you want that, that's fine.  But that' fundamentally not at all the direction we felt was appropriate or even desirable.  Fundamentally, we wanted a way to describe the actual behavior of the runtime, and not introduce a new disparate domain for types that would not be binary compatible.  Going that path, in our views, would have been a complete non-starter for the ecosystem as:

1. it would have been very expensive to adopt.
2. it would have led to real breaks *within* a codebase.
3. it would lead to new libraries being incompatible with older ones.

With the age of this platform, and the enormous amount of code that exists that depends on current semantics, that was just not a viable path.  That is not at all sweeping under the rug.  Rather, it is valuing continuing compatibility and wanting to not force a schism on the ecosystem (e.g. similar to python2/3).  

The approach for `T?` we took was in line with that.  It maximized compat and allowed expressing in the C# type system a way of talking about about generics that was both accurate and in line with NRT.  

I get from this that you would have preferred something far more extreme for the language and runtime.  But, in line with what i was saying before, the only realistic path for that would be full-break, with people either staying behind with the previous runtime (and libraries built for it), and people adopting this new non-compat approach to null with ref types and value types.

**Reply by TahirAhmadov** [reply]
> Yes, this is exactly: if you manage to solve the entire problem here, then the problem is solved. THe point is that we have no known way to go about solving this effectively.

The `Null<T>` I created works. The gaps I'm referring to are the inconvenience aspects of having to do things manually which the language would emit automatically.

> The runtime doesn't base it on size. That's not how the jit works here.

Forget `Null<>` - take any generic type, like `List<Lazy<string>>` and `List<Lazy<object>>` - is it the same problem here?

> I mean, it doesn't break anything in the sense that it would have new semantics as you define them. But it would break htings in that `List<string>` and `List<string?>` would not be the same types, which would be highly undesirable (and basically go against entirely the direction we went with NRT).
> 
> If you're suggesting that we take a generics path with null such that `string` and `string?` are not binary compatible, then that woudl be a complete non-starter on its own.

Yes, that edge case would require a little more work, but it's not crazy. For example:
```cs
D<string?>.List.AddRange(C<string>.List); // AddRange takes IEnumerable, which is covariant, 
// and we add covariance from string to Null<string>
```
In short, I understand that once you start seriously working on a feature, weird things come up. However, you have to weigh the counter-points, as well. For each weird `List<Null<string>>` situation, there is an equally bad, or even worse, `T? FirstOrNull<T>` situation which now cannot be implemented at all.

**Reply by TahirAhmadov** [reply]
To clarify, I'm not saying that `D<string>.List` should be of type `List<string?>`. Those would be different types; and outside of the context of `D<T>`, it would be written out as `List<Null<string>>` - specifically to avoid the confusion. Users would learn that behavior, and how to deal with these details.

**Reply by CyrusNajmabadi** [reply]
> The Null<T> I created works. 

It works for *your* use cases.  For me, if i ended up with `List<string?>`  in one place being incompatible with `List<string?>` in another that woudl be very bad.  Let's use a case that uses your semantics:

```c#
class A<T> where T : class
{
     public static List<T?> List;
}

// This is your new semantics.
class TA<T>
{
    // equivalent to List<Null<T>>
    public static List<T?> List; 
}

// what are the types of 'a' and 'b' ehre?  are they binary compatible?
var a = A<string>.List;
var b = TA<string>.List;
```

?

**Reply by CyrusNajmabadi** [reply]
> In short, I understand that once you start seriously working on a feature, weird things come up. However, you have to weigh the counter-points, as well. 

You're acting like we didnt weigh that.  *We did*.  I'm not sure how to say that more clearly.  All paths had non-ideal parts.  We picked the path that we felt was the *least* non-ideal, and the *most* ideal.  You don't like our assessment.  That's fine.  But painting it as if we swept this under the rug is BS.  We discussed and debated this for *years*.  We experimented and assessed, and we learned and revised.  Eventually we made a call based on all of that.

**Reply by CyrusNajmabadi** [reply]
> take any generic type, like List<Lazy<string>> and List<Lazy<object>> - is it the same problem here?

It is not.  Those are all reference types.

**Reply by TahirAhmadov** [reply]
>     1. it would have been very expensive to adopt.

No argument there, and I respect you trying to prioritize things. No problem.

>     2. it would have led to real breaks _within_ a codebase.
>     3. it would lead to new libraries being incompatible with older ones.

This, I don't think so. With `T?` becoming `Null<T>`, new code using this feature would be aware of it's best-practices. Same thing with `Task<T>` - developers (over time) learned how to work with `await`, etc., to get what I need. I don't view `Null<T>` usability to be any worse than that of `Task<T>`.

Once more, maybe I shouldn't have said "sweep under the rug". But it's a constant annoyance for me in my codebase - I have to write things like `FirstOrNull` and `FirstOrNullStruct`; dictionary `GetOrNull` and `GetOrNullStruct`; some UI controls which I described in my last post under #4080; etc. My point is, the decision that was made, causes me real harm; whereas the `List<Null<string>>` case you mentioned - I can't even imagine getting something like that in the real world.

**Reply by CyrusNajmabadi** [reply]
> Yes, that edge case would require a little more work, 

You keep on handwaving this.  The owners of these consider these incredibly massive pieces of work, and that's based on actual explorations into these spaces.  Again, your position is: if you just make it work, it woudl work, and i think it woudl be not a challenge to do that.

At this point, i'm not sure what to say.  I've explained even on a cursory glance how these things break down, and the people who are experts in the respective domains on what it would take to truly deeply implement this say that it's both enormously expensive *and* incompatible with prior versions of the runtime and libs.   You want us to ahve taken that route.  We chose not to based on what we felt were unacceptable impacts on the ecosystem.

**Reply by CyrusNajmabadi** [reply]
> To clarify, I'm not saying that D<string>.List should be of type List<string?>. Those would be different types;

ANd this would be a deal-breaker for us.   I get it isn't for you.  But that's an irreconcilable difference that we can't cross.  This would split the binary abi for reference types, generics, and null/nullable.  It's not acceptable to us to go that route.

**Reply by TahirAhmadov** [reply]
> ```cs
> class A<T> where T : class
> {
>      public static List<T?> List;
> }
> 
> // This is your new semantics.
> class TA<T>
> {
>     // equivalent to List<Null<T>>
>     public static List<T?> List; 
> }
> 
> // what are the types of 'a' and 'b' ehre?  are they binary compatible?
> var a = A<string>.List;
> var b = TA<string>.List;
> ```

No, they are not. However, I don't think that's a big problem. Like I said, with variance and implicit conversions, most of that pain goes away. It can be a problem if `b` is cast to `object`, then cast to `List<string?>`, but that's already kind of a hole in NRT system. For example, you can also cast `List<string?>` to `object`, then cast that to `List<string>`, and if the developer relies on NRTs, he can get NREs. I don't view it as a strong reason to avoid the feature.

**Reply by CyrusNajmabadi** [reply]
> This, I don't think so. With T? becoming Null<T>, new code using this feature would be aware of it's best-practices

I have no idea waht that means.  Nor do i accept it as a best practice.  Indeed, it would fail at one of the very basic things we want of `?`, the ability to annotate a codebase accurately to understand true runtime null-semantics, without having any runtime/abi impact.

> Same thing with Task<T>

Again, you are saying: it's ok to have an ABI/runtime break with code that updates to use `?` because i think it's ok to have an ABI/runtime break.  We do *not* think it's ok.  That's an irreconcilable difference in beliefs here.  Repeating that you think it's ok doens't change the equation for us because all hte information we based our decisions on remains the same.

>  But it's a constant annoyance for me in my codebase 

Sure.  And we absolutely discussed and debated that.  Many people want some way to make this work.  But the alternative was way more than "constant annoyance".  In comparison, we went with what we felt was the least painful approach.  You're in the group that finds that unpalatable.

> My point is, the decision that was made, causes me real harm; whereas the List<Null<string>> case you mentioned - I can't even imagine getting something like that in the real world.

Yes.  *you* can't.  I really dont' know how much more i can repeat this.  For some users the POR we took would be less ideal.  But in aggregate we felt that was less bad versus the pain that would be caused by going with approaches like you mentioned.  Forcing even *tiny* abi issues was already something we found was a dealbreaker for large parts of the ecosystem.  A literal deal-breaker.  Like no upgrade.  No usage.  No willingness at all to even contemplate adopting these feature areas due to teh risk.  That's the side we have to balance against.  While you cannot envision running into these cases, many groups and organizations absolutely did and just found that the risk was completely unacceptable.  We have to balance that that group of users and organizations is just as important as you are.

**Reply by CyrusNajmabadi** [reply]
>  I don't view it as a strong reason to avoid the feature.

Binary/type compat is important for many users and many domains.  I'm not sure what to tell you.  No one is claiming it's important for you or that you wouldn't be ok if we'd gone your route.

**Reply by TahirAhmadov** [reply]
> > This, I don't think so. With T? becoming Null, new code using this feature would be aware of it's best-practices
> 
> I have no idea waht that means. Nor do i accept it as a best practice. Indeed, it would fail at one of the very basic things we want of `?`, the ability to annotate a codebase accurately to understand true runtime null-semantics, without having any runtime/abi impact.

That's a valid point; inside a context of unconstrained `T` generic, `T?` means something different from what `string?` means, and what `T: class T?` means. However, we already have this divergence with `where T : class` and `where T : struct` - `T?` means totally different things there, as well. Maybe that's an argument for a different symbol? 

> > Same thing with Task
> 
> Again, you are saying: it's ok to have an ABI/runtime break with code that updates to use `?` because i think it's ok to have an ABI/runtime break.

No no, the last thing I want is a break. The `Null<T>` would be a new feature; just like `Task<T>` was a new feature. New libraries returning `Task<T>` are not directly compatible with old libraries which expect `T`. That didn't stop us from adding `Task<T>`, did it?

> > My point is, the decision that was made, causes me real harm; whereas the List<Null> case you mentioned - I can't even imagine getting something like that in the real world.
> 
> Yes. _you_ can't. ... many groups and organizations absolutely did and just found that the risk was completely unacceptable. 

And here, I completely respect your position - different groups have different needs and you have to balance it. I just wish I was part of the conversation early on and saw the concrete examples of issues those people would have faced with my approach. Water under the bridge now...

**Reply by CyrusNajmabadi** [reply]
>  Maybe that's an argument for a different symbol?

Indeed :)

**Reply by CyrusNajmabadi** [reply]
> No no, the last thing I want is a break. The Null<T> would be a new feature; just like Task<T> was a new feature. New libraries returning Task<T> are not directly compatible with old libraries which expect T. That didn't stop us from adding Task<T>, did it?

`Task<T>` isn't T, clearly.  We've never even tried to imply differently.  But, as i've mentioned a bunch now, it was imperative that `T?` be `T` for reference types.  It was literally a deal breaker for lots of the ecosystem for it not to be.  And given the choice for unconstrained generics of having it now mean something new (and ABI incompatible), or having it represent NRT for generics, the latter was deemed an absolutely superior choice.  The former was a ton of new complexity for niche cases of people playing in spaces like yours.  The latter was a very valuable way of annotated the tons of code that *actually exists today* to inform the NRT system about what semantic it has to find and fix real issues in shipping code.

You're welcome to ask for some *new* feature which is a brand new, non-abi compatible, reference-type-esque thing for these cases.  However, we've considered it in the past and felt that hte complexity to the language (and runtime) is too high for the narrow case.  We'd much rather invest that time and effort into things like roles and whatnot.

**Reply by TahirAhmadov** [reply]
>  it was imperative that T? be T for reference types.

I 100% agree with that decision.

> The former was a ton of new complexity for niche cases of people playing in spaces like yours.

That's true and false, IMO. Yes, I'm not entitled to any greater amount of your time than other users. But, because it's limited to some types and helper methods, like my UI control which can have a value of either struct or class type, or the LINQ extensions etc., it doesn't really have that much of an impact on everybody else. For most people, they would get new methods like `FirstOrNull`, which return `Null<T>`, and in the consumption side, it's auto-converted to either `T?` for classes or the existing `Nullable<T>` for structs. That, I think, would be welcomed by everybody as an improvement.

**Reply by HaloFour** [reply]
With something like `Null<T>` you have two unsavory choices:

1. Break the existing ecosystem by updating libraries so that they pass `Null<T>`.
2. Have to build a second shadow ecosystem where those libraries are correctly updated to use `Null<T>`, but all of the existing libraries are still around with their `T` that allow `null`.

Either way you end up with a schism that will undoubtedly take decades to resolve, if ever.

**Reply by TahirAhmadov** [reply]
@HaloFour I thought about that. I think for LINQ, new methods can be created; existing ones would have been decorated with NRT attributes to achieve a similar effect. In other cases, however, the unconstrained `T?` is a brand new feature - there would be nothing to update, it's all new development. And if a new library returns `Null<T>`, and an existing one expect `T?`, and `T` is class, then my proposal accounts for it - it would be auto-converted ("unwrapped"), as needed, and vice-versa. I think people are over-exaggerating the potential problems with it.

**Reply by CyrusNajmabadi** [reply]
? nd if a new library returns Null<T>, and an existing one expect T?, and T is class, then my proposal accounts for it - it would be auto-converted ("unwrapped"), 

There is no way to do this unwrapping at the callsite.  recall that the consumption code might be on prior versions of .net or C#.  

> I think people are over-exaggerating the potential problems with it.

Literally one of the first things i might ever write (statically cached data) is not binary compatible.  I would hit that in one minute of trying to use this.  That's a serious and significant problem.  

>  In other cases, however, the unconstrained T? is a brand new feature 

No, it really wouldn't be.  And that's a core problem with this perspective here.  By using `T?` you are dropping into the NRT world, and behaving in a way that is now incompatible with the ideas that we firmly imbued into that world woudl be a serious problem.  Rember that a prime reason we even did *anything* in this space was because of the high feedback and problems people were having *annotating* real generic code to be NRT safe.  That was much more of an issue than people in your position wanting a uniform representatin of null over value-types and reference-types.  

You are asking us to use the same syntax to solve a *different* problem, and at hte same cause a schism with the group that very much needed help using this syntax to improve the NRT+generics scenario.  I get you think your use case was more important, and you should have gotten first dips on `?`.  We disagree.  I'm not sure what else can be said about that.

If you want your use case to be reconsidered, that's totally fine.  Create a reasonable proposal with reasonable syntax that works well with the current form of C# and doesn't clash with it.  If you do, there's a potential future for it.  But lamenting that we decided to take a piece of syntax to mean something that we think is the superior interpretation for the ecosystem isn't going to go anywhere.  We're certainly not going to change the meaning of it, and we still think it was the right choice given all the constraints we're operating under.

**Reply by sab39** [reply]
I know that in concrete terms the question is moot, and maybe that means that even considering the hypothetical is unproductive, but I feel like it might help with reconciling the conflicting views here: @CyrusNajmabadi, would you agree that *if* we could go back to 2005 (or 2004 or whenever the implementation of C#2 started), and design from scratch without `Nullable<T>` existing yet, then knowing everything we know now about this topic,
* it would be a no-brainer to want to design nullable value types in a way such that this *wouldn't* be impossible later, and
* that if it *weren't* impossible, it would be a no-brainer for NRT to to be designed such that unconstrained `T?` would mean that `null` was allowable for both value and reference types?

**Reply by CyrusNajmabadi** [reply]
If we were doing it all over again, we'd have a complete story for if `null` even exists and what that concept (or anything related to it) would mean for all of .net.  

ALl this stems from allowing null to exist in the first place, and then trying to reconcile very disparate worlds in .net and what it means in all of that.
